### ISSUE TITLE
Test to verify that ordering of blobs in a namespace is based on priority

### ISSUE THREAD

**Issue by rootulp** [celestiaorg/celestia-app#3164]
## Context

> The order of blobs in a namespace is dictated by the priority of the PFBs that paid for the blob. A PFB with greater priority will have all blobs in that namespace strictly before a PFB with less priority. Priority is determined by the gas-price of the transaction (fee/gas).

From https://celestiaorg.github.io/celestia-app/specs/data_square_layout.html#ordering

## Question

Is this actually enforced? It would be nice to have a test that verifies this behavior.

**Comment by rootulp** [comment]
I think this is actually enforced because if a block proposer rearranges blobs in a namespace then that will result in a data root mismatch in process proposal. So the default layout ordering rules are the only layout ordering rules that will be accepted by the network.

**Comment by rootulp** [comment]
If we wanted the default layout ordering rules to only be defaults and allow other valid layouts, I think block proposers would need a mechanism to pass information about the layout (i.e. blob starting share indexes) in their proposal.

**Comment by evan-forbes** [comment]
> if a block proposer rearranges blobs in a namespace then that will result in a data root mismatch in process proposal.

yeah exactly this. the order of the protobuf encoded block determines the order in the square. the commitment to the protobuf encoded block is in the partset header. I think we used to have an exact test for this before refactoring things into blobtxs, but this one is similar but not the same since blobTx are kept inside the block https://github.com/celestiaorg/celestia-app/blob/80968ccea7daf8cfe9d062dc618f8e3e8f783bd9/app/test/process_proposal_test.go#L236-L244

we could add one tho

**Comment by rootulp** [comment]
## Investigation Findings

I investigated how blob ordering within a namespace works end-to-end across celestia-core and celestia-app. Here's what I found:

### 1. The CAT mempool DOES return transactions in priority order

The `t.Skip("TODO: skipping test until prioritisation available in core")` in `app/test/priority_test.go` is outdated. celestia-core v0.39.25's CAT mempool fully implements priority-based ordering:

- `ReapMaxBytesMaxGas` explicitly returns transactions "ordered by decreasing priority, with ties broken by increasing order of arrival"
- The store maintains `orderedTxSets` sorted by aggregated priority via binary search insertion
- `state/execution.go:156` calls `ReapMaxBytesMaxGas` and passes the result to `PrepareProposal`

### 2. Priority ordering is per-sender-group, NOT strict per-transaction

Transactions are grouped into `txSet`s by signer. Each set uses a **gas-weighted average priority**: `aggregatedPriority = Σ(priority_i × gasWanted_i) / Σ(gasWanted_i)`. All transactions from one sender are emitted together (in sequence order) before moving to the next sender's set.

This creates an edge case: if sender Alice has txs with priorities [100, 1], her aggregated priority is 50. If sender Bob has a single tx with priority 60, Bob's tx is reaped first — even though Alice has a tx with priority 100.

**Example:**
| Sender | Tx | Individual Priority | Sender Aggregate |
|--------|------|---------------------|------------------|
| Alice | tx1 | 100 | (100+1)/2 = **50** |
| Alice | tx2 | 1 | |
| Bob | tx1 | 60 | **60** |

**Reap order:** Bob:tx1(60), Alice:tx1(100), Alice:tx2(1) — priority 60 comes before priority 100.

### 3. PrepareProposal preserves ordering but doesn't enforce it

`FilteredSquareBuilder.Fill()` processes blob transactions in whatever order it receives them from the mempool. It does NOT re-sort by priority. The go-square builder then sorts blobs by namespace while preserving input order within each namespace.

### 4. ProcessProposal enforces deterministic square construction

`ProcessProposal` reconstructs the square deterministically and compares data roots. If blobs are reordered, the data root mismatches and the proposal is rejected. This means the only valid ordering is the one produced by the deterministic square construction algorithm.

### 5. The spec is inaccurate

`data_square_layout.md` states: *"A PFB with greater priority will have all blobs in that namespace **strictly** before a PFB with less priority."* This is misleading because the ordering is by sender-aggregated priority, not strict per-transaction priority.

---

## Concrete Steps

### celestia-core PR
- Add `mempool/cat/priority_ordering_test.go` with table-driven tests that:
  - Verify unique-sender txs are reaped in strict priority order (7 test cases)
  - **Document the aggregation edge case** with tests proving same-sender grouping breaks strict per-tx ordering (3 test cases)
  - Stress test with 200 random-priority txs
  - Verify same-sender txs are grouped and ordered by sequence

### celestia-app PR
- **Update `specs/src/data_square_layout.md`**: Replace the inaccurate "strictly before" language with accurate description of sender-aggregated priority ordering
- **Update `specs/src/block_proposer.md`**: Clarify that "priority order" means mempool-provided order (sender-aggregated), not strict per-tx priority
- **Add `app/test/blob_ordering_test.go`**: Tests that PrepareProposal preserves input ordering and ProcessProposal rejects reordered proposals
- **Remove `t.Skip` from `app/test/priority_test.go`**: Prioritization is available in celestia-core v0.39.25

Closes #3164
